---
title: "Usage"
---

# Requirements

1. Wrapper script needs at least **Python 2.7** in order to run. 
2. **[mrfast](http://mrfast.sourceforge.net/) and [mrsfast](http://mrsfast.sourceforge.net/Home)** should be located within the `PATH` in order for wrapper script to complete.  
3. Binary has been compiled on **CentOS 5.x with gcc 4.7 & glibc++ 3.4**, and it might not work with older distributions or versions of glibc++.

### Caveats

While genotyping is usually very fast, mapping can take a lot of time.
In addition to this, mrfast (unlike mrsfast) is known to load all of the reads in memory, so it can create problems for a sample with large FASTQ files. It is recommended that you split large FASTQ files to smaller ones (we usually use 20,000,000 lines per FASTQ file in our test configuration).

The provided script assumes that you will invoke the mapping on only one (small) FASTQ file. You can pass `--sam [mymap.sam]` parameter to the wrapper script instead of `--fastq` parameter if you want to map the FASTQ files manually. Please note that both `mymap.sam` and `mymap.sam.paired` need to be present for Cypiripi in order to produce correct results. 

Lines 123--137 in `cypiripi.py` contain the commands and parameters used for generation of necessary SAM files

# Usage of Cypiripi binary

<div id="arguments" markdown="1">

### Parameter explanation

- `-f [reference file]`

	Pre-processed gene reference file. Use `reference.combined.align` provided in the package for CYP2D6/CYP2D7 gene.

- `-s [SAM file]`

	Input SAM (mapping) file.

- `-p [paired SAM file]`

	Specify the input SAM file containing the paired-end mappings.

	> **Note**: For every read, its pair mapping should appear in the line after read's mapping (e.g. lines should be as R1/1, R1/2, R2/1, R2/2 ...). Do not sort your mapping files by the mapping coordinate, otherwise Cypiripi will fail. This is done by default by mrfast and mrsfast.

	Default: **SAM file with `.paired` suffix**

- `-E [exclusions]`

	A file containing the read names of the reads to be excluded (e.g. CYP2D8-originating reads).

	Default: **SAM file with `.discard` suffix**

- `-C [coverage]`

	Expected coverage per chromosome for sample.

	Default: **20**

- `-T [threshold]`

	Expected minimum cut-off threshold. We recommend 25% or 30% of the coverage.
	
</div>

